lib/queries/community.ts

```diff
@@ -0,0 +1,158 @@
+      CommunityInfo[],
+      Error,
+      InfiniteData<CommunityInfo[]>,
+      (string | number)[],
+      number // pageParam type
+    >,
+    | "queryKey"
+    | "queryFn"
+    | "getNextPageParam"
+    | "initialPageParam"
+    | "getPreviousPageParam"
+  >,
+) => {
+  const limitInt = Math.floor(limit);
+
+  return infiniteQueryOptions({
+    initialPageParam: 0,
+    queryKey: ["communities", limitInt],
+    queryFn: async ({ pageParam = 0 }) => {
+      const client = new GnoJSONRPCProvider(
+        process.env.NEXT_PUBLIC_ZENAO_GNO_ENDPOINT || "",
+      );
+      const res = await client.evaluateExpression(
+        `gno.land/r/zenao/communityreg`,
+        `communitiesToJSON(listCommunities(${limitInt}, ${pageParam * limitInt}))`,
+      );
+      const raw = extractGnoJSONResponse(res);
+      const json = communitiesListFromJson(raw);
+
+      return json;
+    },
+    getNextPageParam: (lastPage, pages) => {
+      if (lastPage.length < limitInt) {
+        return undefined;
+      }
+      return pages.length;
+    },
+    getPreviousPageParam: (firstPage, pages) => {
+      if (firstPage.length < limitInt) {
+        return undefined;
+      }
+      return pages.length - 2;
+    },
+    ...options,
+  });
+};
+
+export const communityUserRoles = (
+  communityId: string | null,
+  userAddress: string | null,
+) =>
+  queryOptions({
+    queryKey: ["communityUserRoles", communityId, userAddress],
+    queryFn: async () => {
+      if (!communityId || !userAddress) {
+        return [];
+      }
+
+      const client = new GnoJSONRPCProvider(
+        process.env.NEXT_PUBLIC_ZENAO_GNO_ENDPOINT || "",
+      );
+      const res = await client.evaluateExpression(
+        `gno.land/r/zenao/communities/c${communityId}`,
+        `community.GetUserRolesJSON(${JSON.stringify(userAddress)})`,
+      );
+      const roles = extractGnoJSONResponse(res);
+      return communityGetUserRolesSchema.parse(roles);
+    },
+  });
+
+export const communityUsersWithRoles = (
+  communityId: string,
+  roles: CommunityUserRole[],
+) =>
+  queryOptions({
+    queryKey: ["communityRoles", communityId, JSON.stringify(roles)],
+    queryFn: async () => {
+      const client = new GnoJSONRPCProvider(
+        process.env.NEXT_PUBLIC_ZENAO_GNO_ENDPOINT || "",
+      );
+      const res = await client.evaluateExpression(
+        `gno.land/r/zenao/communities/c${communityId}`,
+        `community.GetUsersWithRolesJSON(${jsonStringArrayIntoGoArray(roles)})`,
+      );
+      const raw = extractGnoJSONResponse(res);
+
+      return communityUsersWithRolesResponseSchema.array().parse(raw);
+    },
+  });
+
+function communitiesListFromJson(raw: unknown) {
+  const list = raw as unknown[];
+  return list.map((elem) =>
+    fromJson(CommunityInfoSchema, elem as CommunityInfoJson),
+  );
+}
+
+function jsonStringArrayIntoGoArray(arr: string[]): string {
+  return `[]string${JSON.stringify(arr).replace("[", "{").replace("]", "}")}`;
+}
```

Comment:
```suggestion
function goStringSliceLiteral(arr: string[]): string {
  return `[]string{${arr.map(s => JSON.stringify(s)).join(",")}}`;
}
```
- this is called a string slice literal
- the version with replace is very brittle, it will modify content when the strings contains `[` or `]`